deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java

```diff
@@ -144,17 +144,50 @@ public DeployBean load(String deployId) throws Exception {
 
     // Keep host and group membership in sync
     void updateHosts(String hostName, String hostIp, String hostId, Set<String> groups) throws Exception {
-        hostDAO.insertOrUpdate(hostName, hostIp, hostId, HostState.ACTIVE.toString(), groups);
-
-        List<String> recordedGroups = hostDAO.getGroupNamesByHost(hostName);
+        Set<String> recordedGroups = new HashSet<String>(hostDAO.getGroupNamesByHost(hostName));
+
+        Set<String> groupsToAdd = new HashSet<String>();
+        // Insert if not recorded
+        for (String group : groups) {
+            if (!recordedGroups.contains(group)) {
+                LOG.info("Insert host {} to group {} mapping", hostName, group);
+                groupsToAdd.add(group);
+            }
+        }
+        if (groupsToAdd.size() > 0) {
+            hostDAO.insertOrUpdate(hostName, hostIp, hostId, HostState.ACTIVE.toString(), groups);
+        }
+        
+        // Remove if not reported
         for (String recordedGroup : recordedGroups) {
             if (!groups.contains(recordedGroup)) {
-                LOG.warn("Remove host {} from group {}", hostName, recordedGroup);
+                LOG.warn("Remove host {} to group {} mapping", hostName, recordedGroup);
                 this.hostDAO.removeHostFromGroup(hostId, recordedGroup);
             }
         }
+    }
 
-
+    void updateHostStatus(String hostId, String hostName, String hostIp, String agentVersion, String asg, Set<String> groups) throws Exception {
+        HostAgentBean hostAgentBean = hostAgentDAO.getHostById(hostId);
+        if (hostAgentBean == null) {
```

Comment:
I prefer to split this function into 2 more to improve code readness. 
This isn't a requirement we can discuss this if want.